docs/uprint/uprint-lpr.1.pod
#
# mouse:~ppr/src/docs/uprint/uprint-lpr.1.pod
# Copyright 1995--2004, Trinity College Computing Center.
# Written by David Chappell.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
# 
# * Redistributions of source code must retain the above copyright notice,
# this list of conditions and the following disclaimer.
# 
# * Redistributions in binary form must reproduce the above copyright
# notice, this list of conditions and the following disclaimer in the
# documentation and/or other materials provided with the distribution.
# 
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS BE 
# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR 
# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF 
# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE 
# POSSIBILITY OF SUCH DAMAGE.
#
# Last modified 11 December 2004.
#

=head1 NAME

uprint-lpr - selects spooler and submits print job

=head1 SYNOPSIS

B<uprint-lpr> I<switches> [I<filename1> [I<filename2>] ... ]

=head1 DESCRIPTION

This program is part of the PPR uprint suite.  See the UPRINT(1) man page
for an explanation of what this means.

Your system administrator may have choosen to replace F</usr/bin/lpr>
with a link to this program.

This program is designed to be a drop-in replacement for the B<lpr> from BSD Unix.
It also accepts some of the options accepted by enhanced versions of B<lpr> found
on certain other Unix systems such as DEC OSF.  It follows traditional B<lpr> semantics
as closely as possible.  

These are the option that B<uprint-lpr> accepts:

=over 4

=item B<-P> I<queuename>

Use this option to select a print queue.  If this option is omited, then
the queue name is read from the environment variable PRINTER.  If PRINTER
is not defined then the queue name will be read from the "uprint-lpr=" entry
in the "[default destinations] section of F<uprint.conf>.  If that too is
missing then the queue be "lp".

Note that B<uprint-lpr> without the B<-P> option may choose
a different queue that B<ppr> or B<uprint-lp> without the B<-d> option would.

=item B<-m>

Send email when the job is done.

=item B<-#> I<n>

Print the indicated number of copies.

=back

=head2 Input File Type Options

=over 4

=item B<-p>

The files are plain ASCII text which should be passed through B<pr> before
printing.

=item B<-l>

The files are in some sort of line printer format.  Control characters should
be preserved and no attempt should be made to insert page breaks.

=item B<-f>

The files use FORTRAN carriage control.

=item B<-c>

The files are cifplot(1) output.

=item B<-g>

The files are plot(1) output.

=item B<-d>

The files are TeX DVI output.

=item B<-n>

The files are Ditroff (modern Troff) output.

=item B<-t>

The files are old Troff (CAT/4 typesetter format) output.

=item B<-v>

The files are raster images suitable for the "Benson Varian", whatever
that is.

=back

=head2 Banner Page Options

=over 4

=item B<-h>

Don't print a banner page.

=item B<-C> I<string>

Set the job classification to be printed on the banner page.  If this switch
is ommited then the name of the computer from which the job origionated will
be used instead.  (Exactly what a "job classifcation" is is unclear.)

=item B<-J> I<string>

Set the job name for queue listings and banner pages.

=item B<-T> I<string>

Select the title to appear in the running heads created by B<pr>.  (Obviously,
or not so obviously, this switch only makes a difference in combination
with the B<-p> switch.)

=back

=head2 Other Standard Options

=over 4

=item B<-r>

Delete the job when printing is done.

=item B<-s>

Don't make a copy of the files, just make symbolic links to them.  If you use
this then you probably shouldn't modify the files until printing is done.
In many cases, Uprint will ignore this option and make its own copy of the
files anyway.

=item B<-1 I<fontname>>

=item B<-2 I<fontname>>

=item B<-3 I<fontname>>

=item B<-4 I<fontname>>

Select fonts to be mounted in Troff font positions 1 through 4.

=item B<-i> I<n>

Specifies a number of blanks as a left margin.  The default is supposedly 8.
It is unclear when this switch is supposed to be effective.  It probably
will have an effect when printing plain ASCII text (no file type switch)
and when using the B<-p> switch.

=item B<-w> I<n>

This specifies the page width (in columns) for B<pr>.

=item B<-U> I<user>

Supposedly, this switch is used specify the owner of this job.  Only certain
users can use it.  It is used by filters that requeue jobs.  At least that
is what the BSD lpr(1) man page says.  Likely, this switch does not work
correctly in B<uprint-lpr>.

=back

=head2 OSF/1 and Digital Unix Compatibility Options

The B<lpr> program in OSF/1 and Digital Unix has additional options.
B<Uprint-lpr> understands these additional options.

=over 4

=item B<-I> I<trayname>

Specifies the input tray name.  Valid values depend on the queue.  For PPR
queues this will be the tray name from the PPD file.  There will probably
not be any warning if you select a non-existent tray.

=item B<-K> I<duplexmode>

This option selects double sided printing.  It can also be used to select
printing modes which are useful when the printing will be done on one
side but will be photocopied onto two sides.  Here are the options:

The various duplex modes control three things.  On is how many sides of the
page are used.  The other is, if both sides are used, how is the page flipped.
The third thing is page layout.

Page layout often changes according to how a document will be bound.
When a document is going to be bound
along one edge it is customary to make that margin a little wider.  If a
document is printed on one side only and bound along the left hand edge,
the extra space is added to the left margin.  If it is printed on two sides
and bound on the left hand edge then the extra space is added to the left
margin on odd numbered pages and to the right margin on even numbered pages.
This extra space is sometimes called a gutter.  Also, if page numbers are
printed and they are printed off center then when printing on two sides they
are generall placed on the right on odd numbered pages and on the left on
even numbered pages.

=over 4

=item<one_sided> or B<1> or B<one_sided_simplex>

This is the default.  Page are printed on one side.  Where the spooler can
adjust margins, it will put any "gutter" (extra space for the bound edge)
on the left hand side.

=item B<two_sided> or B<2> or B<two_sided_duplex>

The document is printed on both sides of the paper.  This is probabaly
the duplex option you want.  The pages are flipped over the long edge
(generally the left edge, like most books).  If a gutter is used,
it will alternate between the left and right margins.

=item B<tumble> or B<two_sided_tumble>

The document is printed on both sides of the paper.  The pages
are flipped over the short edge (generally the top edge, like
most calendars).  If a gutter is used, it will alternate between
the top and buttom margins.

=item B<one_sided_duplex>

The document is printed on one side of the paper only, but if there
is a gutter it alternates betwen the left and right margins.  The idea is
that the finished pages can then be photocopied onto both sides.

=item B<one_sided_tumble>

This is like B<one_sided_duplex> but the gutter alternates between the top
and bottom margins.

=item B<two_sided_simplex>

The document is printed on both sides of the paper (duplexed) but the
gutter is always added to the left margin.  It is unclear why you might want
to do this.  The option is probably present only for the sake of completeness.

=back

=item B<-O> I<orientation>

This selects the orientation of the output.  The choices are "portrait"
and "landscape".  In the case of a PPR queue this parameter will only make
a difference if the input file must be converted to PostScript, and then
only if the filter understands the "orientation=" option.

=item B<-j>

Print the jobid.

=item B<-N> I<n>

Print the pages N-Up.

=item B<-o> I<n>

=item B<-o> I<tray>

This option specifies the output tray.  The Digital Unix lpr(1) man page says
that the parameter should be a number, but for a PPR queue it should be an
output tray name.  The output tray name should be one of the machine
readable names mentioned in the PPD file.  If you select a non-existent
output tray there will be no warning
and the job will probably be sent to an output tray you did not intend.

=item B<-x>

Indicates that the file does not require filtering before printing.
This option may not work when printing to PPR queues.

=back

=head1 ENVIRONMENT VARIABLES

To learn how the environment variable PRINTER is used, see the description of the B<-P> option.

The variables PPR_RESPONDER, PPR_RESPONDER_ADDRESS and PPR_RESPONDER_OPTIONS
can be used to control the response method used to inform the submitter of 
the fate of such jobs as are eventually processed by PPR (as opposed to some 
other spooler.  They work as described in the L<ppr.1> manpage.  If the job is
passed to a remove node's PPR spooler through the LPR/LPD protocol and 
B<lprsrv>, a protocol extension is used to convey this information.

=head1 BUGS

Many options are simply passed to the remote spooler without any validity
checking.  This means that incorrect options will often simply be ignored.
Some of the more obscure options may not be implemented correctly.  Bug reports are welcome 
and should be sent to ppr-bugs@trincoll.edu.

=head1 AUTHOR

David Chappell, Trinity College Computing Center, Hartford, Connecticut.

=cut



History.txt
mouse:~ppr/src/History.txt (symbolic link)

This file contains a history of changes to PPR which occured after the
release of version 1.00.


1.00, 26 April 1995

First public release.


1.10, 27 April 1995--7 June 1995

Removed "From:" line from message passed to Sendmail by pprd when sending
a message listing the remaining printer faults.  This is so that Sendmail
will take the "real name" from the password file and put it in the "From:"
line.

Removed dotmatrix/input.epson, a GIF file which had been inadvertently
included in the distribution.

Moved the contents of include/atalk_streams.h into
interfaces/atalk_streams.c since it was not long enough to justify an
include file, nor was it of interest to any other program.

Added code to ppad to classify members of the group "ppradmin" as PPR
addministrators who can edit printer and group configurations.

Added -v switch to ppad.

Removed uid's 1-99 from the list of priveledged users.

Fixed ppad(8) so that it considers the group "ppradmin" as priveledged
rather than "ppr".

Changed ppr(1) so that it will accept the members of the group "pprprox" as
priveledged as well as the members of "ppr".  This is for security reasons
as one may not want to grant a user who has the privedge of using the -f
switch the priveledge of deleting other people's jobs.

Modified ppop(1) and pprd(8) so that user's can request to see the queue
entry for a single job.  Previously, the only options were all jobs for a
certain destination or all jobs in all queues.

Rewrote and renamed the undocumented "ppop qdebug" command to become the new
documented command "ppop details".  (This was done primarily to support
wwwppop, an experimental managment interface for PPR.  Since it was still
incomplete it was not released with this version of PPR.)

Renamed interfaces/atalk_streams.c to atalk_ali.c and
papsrv/papsrv_streams.c to papsrv_ali.c

Removed code from "make dirs" so that it doesn't create the directory
"/etc/ppr/groups.ppd".

Added sleep=N and eoftimeout=N options to tcpip interface.

Added -C _filename_ switch to ppr invokation in xppr.

Added "-default okay" to xmessage invokation in xppr.

Added the "ppop lpq" sub-command to the "ppop help" information, thus
making it a documented sub-command.

Extended automatic bin selection to multiple media types when "%%PageMedia:"
comments are present.  This has not yet been well tested.

Made "%%EndProcset" an acceptable substitute for "%%EndProcSet" due to a
reputed bug in QuickDraw GX.

Changed rule in ppr_infile.c for detecting PCL code so that documents with
only one escape would still be detected.


1.20, 8 June 1995--10 January 1996

Fixed code in ppr_infile.c so that the ppr -o switch will work even if there
is no "DefFiltOpts:" line in the printer or group configuration file.

Added '(' to list of likely second characters for HP escape codes
in ppr_infile.c

Replaced printing of errno value with strerror(errno) output in some places,
eliminated printing of errno=x from other places where both were printed.

Fixed ppr(1) so that it sets the environment variables "PATH", "IFS", and
"SHELL" so that filters and responders it invokes are less easy to subvert.

Removed setting of PATH from ~ppr/src/misc_filters/tex.perl and changed
shell name from "/usr/bin/tainperl" to "/usr/bin/perl".

Changed code which determins default responder address so that the
environment variables "USER" and "LOGNAME" are used in preference to the
current uid.

Added prompts in Configure to set the directory into which the links to PPR
binaries will be written and the path to the X-Windows binaries such as
xterm.

Modified papsrv so that it will pass the network and node number of the
client to ppr with the -r switch.  (This is for use with an experimental
AppleTalk responder.)

Modified papsrv.c, renaming cleanup() to termination_handler() and causing
it to call fatal() only the first time it is called.  Without this, name
could be left on the network if the CAP version received two terminate
signals in rapid sucession.

Added code to re-line-wrap responder messages for certain responders.
(The library function responder_width() returns the width for a responder.
Currently it is hard coded to return zero for all responders except "samba".
A return value of zero indicates that the message should not be
re-line-wrapped.)

Dropt reading of LPDEST and PRINTER by ppr(1).  (I can not really remember
why I put this feature in.  I don't think the reasons where very good.  It
may have been that I thought that the same printer would be available thru
two spooling systems so one might want to set the default destination for
both with one variable.  That now seems a little far fetched.)

Separated Solaris and Starserver sections of sysdep.h

Added PNG (Portable Network Graphics) format with magic number recognition
to ppr_infile.c.  This remains untested since I do not have any programs
which manipulate this graphics format.

In pprd.c, put sleep(1) between execl("pprdrv") and _exit(242) so that pprd
has time to record the PID before the child exits.  (If this is not done
and execl() fails immediately, a child which should have become pprdrv may
exit before the parent records its PID in the information structure for that
queue.  This result is that it will not recoginize the purpose of the child
when it receives notification of its death and will leave the queue in the
printing state indefinitely.)

Did same in pprdrv.c in code which executes interface.  Also changed
_exit(255) to _exit(242).  (Exit 242 is the new standard since PERL uses 255
as the exit code when die() is executed.)

Added automatic creation of "DefFiltOpts:" lines to ppad(8).  Manual
creation is supported by the the "ppad deffiltopts" and "ppad group
deffiltopts" command.

Modified ppop so that interactive mode restores the terminal settings when
it exits.  Also changed terminal setting code to use the POSIX method of

uprint/uprint-lpr.c
/*
** mouse:~ppr/src/uprint/uprint-lpr.c
** Copyright 1995--2003, Trinity College Computing Center.
** Written by David Chappell.
**
** Redistribution and use in source and binary forms, with or without
** modification, are permitted provided that the following conditions are met:
**
** * Redistributions of source code must retain the above copyright notice,
** this list of conditions and the following disclaimer.
** 
** * Redistributions in binary form must reproduce the above copyright
** notice, this list of conditions and the following disclaimer in the
** documentation and/or other materials provided with the distribution.
**
** THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
** AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
** IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
** ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS BE 
** LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR 
** CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF 
** SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
** INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
** CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
** ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE 
** POSSIBILITY OF SUCH DAMAGE.
**
** Last modified 19 February 2003.
*/

#include "before_system.h"
#include <unistd.h>
#include <stdlib.h>
#include <stdarg.h>
#include <pwd.h>
#include <string.h>
#ifdef INTERNATIONAL
#include <locale.h>
#include <libintl.h>
#endif
#include "gu.h"
#include "global_defines.h"
#include "uprint.h"

extern char *optarg;
extern int optind;

static const char *const myname = "uprint-lpr";

void uprint_error_callback(const char *format, ...)
	{
	va_list va;
	fprintf(stderr, "%s: ", myname);
	va_start(va, format);
	vfprintf(stderr, format, va);
	fputc('\n', stderr);
	va_end(va);
	} /* end of uprint_error_callback() */

static const char *option_list =
		"cdfglnptvP:hmrs#:1:2:3:4:C:J:T:U:i:w:" /* BSD */
		"I:jK:N:o:O:x"							/* DEC OSF/1 3.2 */
		;

/* OSF lpr compatible duplex options: */
static struct
	{
	const char *from;
	const char *to;
	} osf_K_xlate[] =
	{
	/* These are the options in queue file format.  The fact that
	   "one" and "two" are acceptable to lpr is undocumented. */
	{"one", "one"},
	{"two", "two"},
	{"tumble", "tumble"},
	{"one_sided_duplex", "one_sided_duplex"},
	{"one_sided_tumble", "one_sided_tumble"},
	{"two_sided_simplex", "two_sided_simplex"},

	/* Other documented forms accepted by OSF lpr. */
	{"1", "one"},
	{"one_sided", "one"},
	{"one_sided_simplex", "one"},
	{"2", "two"},
	{"two_sided", "two"},
	{"two_sided_duplex", "two"},
	{"two_sided_tumble", "tumble"},

	{NULL, NULL}
	};

int main(int argc, char *argv[])
	{
	uid_t uid = getuid();
	void *upr;
	struct passwd *pw;
	int c;
	int x;

	/* Initialize internation messages library. */
	#ifdef INTERNATIONAL
	setlocale(LC_MESSAGES, "");
	bindtextdomain(PACKAGE, LOCALEDIR);
	textdomain(PACKAGE);
	#endif

	/* Trap loops: */
	if(uprint_loop_check(myname) == -1)
		return 1;

	/*
	** Create a uprint object.
	**
	** Save the argument array in it in case the uprint
	** library decides to print the job by executing
	** the real lpr.
	*/
	if((upr = uprint_new("lpr", argc, (const char **)argv)) == (void*)NULL)
		{
		fprintf(stderr, "%s: %s\n", myname, uprint_strerror(uprint_errno));
		uprint_delete(upr);
		return 1;
		}

	/*
	** Set the UID number and name of the user:
	*/
	if( (pw=getpwuid(uid)) == (struct passwd *)NULL)
		{
		fprintf(stderr, _("%s: getpwuid(%ld) failed to find your account\n"), myname, (long int)uid);
		uprint_delete(upr);
		return 1;
		}
	if(pw->pw_name == (char*)NULL)
		{
		fprintf(stderr, "%s: strange getpwuid() error, pw_name is NULL\n", myname);
		uprint_delete(upr);
		return 1;
		}
	uprint_set_user(upr, uid, -1, pw->pw_name);

	/*
	** Set the defaults for a few options:
	*/
	uprint_set_immediate_copy(upr, TRUE);

	/*
	** Parse the switches.  Mostly, we will call uprint
	** member functions.
	*/
	while( (c=getopt(argc, argv, option_list)) != -1 )
		{
		switch(c)
			{
			case 'c':			/* CIF */
			case 'd':			/* DVI */
			case 'f':
			case 'g':
			case 'l':
			case 'n':
			case 'p':
			case 't':
			case 'v':
			case 'x':			/* OSF no filtering */
				uprint_set_content_type_lpr(upr, c);
				break;

			case 'P':
				uprint_set_dest(upr, optarg);
				break;

			case 'h':			/* suppress burse page */
				uprint_set_nobanner(upr, TRUE);
				break;

			case 'm':			/* send mail on completion */
				uprint_set_notify_email(upr, TRUE);
				break;

			case 'r':			/* remove files on completion */
				uprint_set_unlink(upr, TRUE);
				break;

			case 's':			/* use symbolic links */
				uprint_set_immediate_copy(upr, FALSE);
				break;

			case '#':			/* number of copies */
				uprint_set_copies(upr, atoi(optarg));
				break;

			case '1':			/* troff fonts */
				uprint_set_troff_1(upr, optarg);
				break;

			case '2':
				uprint_set_troff_2(upr, optarg);
				break;

			case '3':
				uprint_set_troff_3(upr, optarg);
				break;

			case '4':
				uprint_set_troff_4(upr, optarg);
				break;

			case 'C':			/* class */
				uprint_set_lpr_class(upr, optarg);
				break;

			case 'J':			/* job name */
				uprint_set_jobname(upr, optarg);
				break;

			case 'T':			/* title */
				uprint_set_pr_title(upr, optarg);
				break;

			case 'U':			/* user name */
				uprint_set_user(upr, uid, -1, optarg);	/* !!! */
				break;

			case 'i':			/* indent */
				uprint_set_indent(upr, optarg);
				break;

			case 'w':			/* page width */
				uprint_set_width(upr, optarg);
				break;

			case 'I':			/* OSF input tray */
				uprint_set_osf_LT_inputtray(upr, optarg);
				break;

			case 'j':			/* OSF display job ID */
				uprint_set_show_jobid(upr, TRUE);
				break;

			case 'K':			/* OSF sides (duplex) */
				{
				int x;
				for(x=0; osf_K_xlate[x].from; x++)
					{
					if(strcmp(optarg, osf_K_xlate[x].from) == 0)
						{
						uprint_set_osf_K_duplex(upr, osf_K_xlate[x].to);
						break;
						}
					}
				if(! osf_K_xlate[x].from)
					{
					fprintf(stderr, _("%s: Invalid -K argument\n"), myname);
					return 1;
					}
				break;
				}

			case 'N':			/* OSF N-Up */
				x = atoi(optarg);
				if(x < 1)
					{
					fprintf(stderr, _("%s: Invalid -N argument\n"), myname);
					return 1;
					}
				uprint_set_nup(upr, x);
				break;

			case 'o':			/* OSF output tray */
				uprint_set_osf_GT_outputtray(upr, optarg);
				break;

			case 'O':			/* OSF orientation */
				uprint_set_osf_O_orientation(upr, optarg);
				break;

			default:			/* Illegal option */
				break;			/* ignore as BSD lpr seems to */
			}
		}

	/* If the print destination has not yet been
	   determined, determine it now. */
	if(!uprint_get_dest(upr))
		uprint_set_dest(upr, uprint_default_destinations_lpr());

	/* If there are file names, tell uprint about them: */
	if(argc > optind)
		uprint_set_files(upr, (const char **)&argv[optind]);

	/* Look for a PPR responder we can use. */
	{
	char *p;
	if((p = getenv("PPR_RESPONDER")))
		uprint_set_ppr_responder(upr, p);
	if((p = getenv("PPR_RESPONDER_ADDRESS")))
		uprint_set_ppr_responder_address(upr, p);
	if((p = getenv("PPR_RESPONDER_OPTIONS")))
		uprint_set_ppr_responder_options(upr, p);
	}

	/* Print the job. */
	{
	int return_code;

	if((return_code = uprint_print(upr, TRUE)) >= 0)
		{
		/* Child ran, nothing more to do. */
		}
	/* No match anywhere: */
	else if(uprint_errno == UPE_UNDEST)
		{
		fprintf(stderr, _("%s: Print queue \"%s\" not found.\n"), myname, uprint_get_dest(upr));
		/* This is the code that BSD lpr uses for unknown queue: */
		return_code = 1;
		}
	/* Printing failed, */
	else
		{
		/* An arbitrary value intended to indicated a major failure: */
		return_code = 255;
		}

	/* We are done with the UPRINT structure: */
	uprint_delete(upr);

	return return_code;
	}
	} /* end of main() */

/* end of file */

